File: backend/app/schemas/audit.py
from pydantic import BaseModel, Field, ConfigDict
from typing import Optional, Dict, Any, List
from datetime import datetime
from uuid import UUID


class AuditLogBase(BaseModel):
    """Base audit log schema"""
    action: str = Field(..., description="Action performed")
    resource: str = Field(..., description="Resource type affected")
    resource_id: Optional[UUID] = Field(None, description="ID of specific resource affected")
    details: Optional[Dict[str, Any]] = Field(None, description="Additional action details")
    status: str = Field(default="success", description="Action outcome")


class AuditLogResponse(AuditLogBase):
    """Audit log response schema"""
    id: UUID
    user_id: Optional[UUID] = None
    tenant_id: Optional[UUID] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    request_id: Optional[str] = None
    timestamp: datetime = Field(..., description="When the action occurred")

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_audit_log(cls, audit_log):
        """Create response from AuditLog model"""
        return cls(
            id=audit_log.id,
            user_id=audit_log.user_id,
            tenant_id=audit_log.tenant_id,
            action=audit_log.action,
            resource=audit_log.resource,
            resource_id=audit_log.resource_id,
            details=audit_log.details,
            status=audit_log.status,
            ip_address=audit_log.ip_address,
            user_agent=audit_log.user_agent,
            request_id=audit_log.request_id,
            timestamp=audit_log.created_at,
        )


class AuditLogListResponse(BaseModel):
    """Paginated audit log list response"""
    logs: List[AuditLogResponse]
    total: int = Field(..., description="Total number of logs matching filters")
    limit: int = Field(..., description="Number of logs per page")
    offset: int = Field(..., description="Current page offset")

    model_config = ConfigDict(from_attributes=True)


class AuditLogFilter(BaseModel):
    """Audit log filter parameters"""
    user_id: Optional[UUID] = Field(None, description="Filter by user ID")
    action: Optional[str] = Field(None, description="Filter by action type")
    resource: Optional[str] = Field(None, description="Filter by resource type")
    resource_id: Optional[UUID] = Field(None, description="Filter by specific resource")
    status: Optional[str] = Field(None, description="Filter by status (success/failure/error)")
    start_date: Optional[datetime] = Field(None, description="Filter logs after this date")
    end_date: Optional[datetime] = Field(None, description="Filter logs before this date")
    limit: int = Field(default=100, ge=1, le=1000, description="Number of results per page")
    offset: int = Field(default=0, ge=0, description="Pagination offset")


class UserActivityResponse(BaseModel):
    """User activity summary"""
    user_id: UUID
    total_actions: int = Field(..., description="Total number of actions")
    recent_actions: List[AuditLogResponse] = Field(..., description="Recent activity")
    last_login: Optional[datetime] = Field(None, description="Last login timestamp")
    last_activity: Optional[datetime] = Field(None, description="Last activity timestamp")

    model_config = ConfigDict(from_attributes=True)


class SecurityEventResponse(BaseModel):
    """Security event summary"""
    event_type: str = Field(..., description="Type of security event")
    count: int = Field(..., description="Number of occurrences")
    recent_events: List[AuditLogResponse] = Field(..., description="Recent events of this type")

    model_config = ConfigDict(from_attributes=True)


class AuditStatistics(BaseModel):
    """Audit statistics summary"""
    total_logs: int = Field(..., description="Total number of audit logs")
    total_users: int = Field(..., description="Number of unique users")
    total_actions: int = Field(..., description="Number of unique actions")
    failed_logins_24h: int = Field(..., description="Failed login attempts in last 24 hours")
    successful_logins_24h: int = Field(..., description="Successful logins in last 24 hours")
    actions_by_type: Dict[str, int] = Field(..., description="Count of actions by type")
    actions_by_status: Dict[str, int] = Field(..., description="Count of actions by status")

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/lib/api/audit.ts
import { apiClient } from './client';

export interface AuditLog {
  id: string;
  user_id: string | null;
  tenant_id: string | null;
  action: string;
  resource: string;
  resource_id: string | null;
  details: Record<string, unknown> | null;
  status: string;
  ip_address: string | null;
  user_agent: string | null;
  request_id: string | null;
  timestamp: string;
}

export interface AuditLogListResponse {
  logs: AuditLog[];
  total: number;
  limit: number;
  offset: number;
}

export interface AuditStatistics {
  total_logs: number;
  total_users: number;
  total_actions: number;
  failed_logins_24h: number;
  successful_logins_24h: number;
  actions_by_type: Record<string, number>;
  actions_by_status: Record<string, number>;
}

export interface AuditLogParams {
  skip?: number;
  limit?: number;
  action?: string;
  resource?: string;
  status?: string;
  user_id?: string;
  tenant_id?: string;
  start_date?: string;
  end_date?: string;
  search?: string;
}

export interface ArchiveFile {
  name: string;
  size_bytes: number;
  size_readable: string;
}

export interface ArchiveResult {
  message: string;
  archived: number;
  before_date: string;
  file: ArchiveFile | null;
}

export interface ArchiveInfo {
  name: string;
  size_bytes: number;
  size_readable: string;
  created_at: string;
  total_records: number;
  archived_at: string | null;
  before_date: string | null;
}

export interface ArchiveListResponse {
  archives: ArchiveInfo[];
  total: number;
}

export const auditAPI = {
  /**
   * Get paginated list of audit logs with optional filters
   */
  getAuditLogs: async (params: AuditLogParams = {}): Promise<AuditLogListResponse> => {
    const queryParams = new URLSearchParams();

    if (params.skip !== undefined) queryParams.append('skip', params.skip.toString());
    if (params.limit !== undefined) queryParams.append('limit', params.limit.toString());
    if (params.action) queryParams.append('action', params.action);
    if (params.resource) queryParams.append('resource', params.resource);
    if (params.status) queryParams.append('status', params.status);
    if (params.user_id) queryParams.append('user_id', params.user_id);
    if (params.tenant_id) queryParams.append('tenant_id', params.tenant_id);
    if (params.start_date) queryParams.append('start_date', params.start_date);
    if (params.end_date) queryParams.append('end_date', params.end_date);
    if (params.search) queryParams.append('search', params.search);

    const queryString = queryParams.toString();
    const url = `/admin/audit-logs${queryString ? `?${queryString}` : ''}`;

    return apiClient.get<AuditLogListResponse>(url);
  },

  /**
   * Get audit log statistics
   */
  getStatistics: async (): Promise<AuditStatistics> => {
    return apiClient.get<AuditStatistics>('/admin/audit-logs/statistics');
  },

  /**
   * Get list of unique action types
   */
  getActions: async (): Promise<string[]> => {
    return apiClient.get<string[]>('/admin/audit-logs/actions');
  },

  /**
   * Get list of unique resource types
   */
  getResources: async (): Promise<string[]> => {
    return apiClient.get<string[]>('/admin/audit-logs/resources');
  },

  /**
   * Get single audit log by ID
   */
  getAuditLog: async (logId: string): Promise<AuditLog> => {
    return apiClient.get<AuditLog>(`/admin/audit-logs/${logId}`);
  },

  /**
   * Clear all audit logs (DEV MODE only)
   */
  clearLogs: async (): Promise<{ message: string; deleted: number }> => {
    return apiClient.delete<{ message: string; deleted: number }>('/admin/audit-logs/');
  },

  /**
   * Archive old audit logs before a given date
   * Exports logs to JSON file, then deletes from database
   */
  archiveLogs: async (beforeDate?: string): Promise<ArchiveResult> => {
    const params = beforeDate ? `?before_date=${encodeURIComponent(beforeDate)}` : '';
    return apiClient.post<ArchiveResult>(`/admin/audit-logs/archive${params}`);
  },

  /**
   * List all archived audit log files
   */
  listArchives: async (): Promise<ArchiveListResponse> => {
    return apiClient.get<ArchiveListResponse>('/admin/audit-logs/archives');
  },

  /**
   * Get download URL for an archive file
   */
  getArchiveDownloadUrl: (filename: string): string => {
    return `/api/v1/admin/audit-logs/archives/${encodeURIComponent(filename)}`;
  },

  /**
   * Delete an archive file (DEV MODE only)
   */
  deleteArchive: async (filename: string): Promise<{ message: string }> => {
    return apiClient.delete<{ message: string }>(`/admin/audit-logs/archives/${encodeURIComponent(filename)}`);
  },
};

/**
 * Tenant-scoped audit API — uses the same endpoints as auditAPI.
 * The backend automatically scopes results to the current user's tenant
 * when the caller is not a super admin.
 */
export const tenantAuditAPI = {
  getAuditLogs: async (params: Omit<AuditLogParams, 'tenant_id'> = {}): Promise<AuditLogListResponse> => {
    const queryParams = new URLSearchParams();

    if (params.skip !== undefined) queryParams.append('skip', params.skip.toString());
    if (params.limit !== undefined) queryParams.append('limit', params.limit.toString());
    if (params.action) queryParams.append('action', params.action);
    if (params.resource) queryParams.append('resource', params.resource);
    if (params.status) queryParams.append('status', params.status);
    if (params.user_id) queryParams.append('user_id', params.user_id);
    if (params.start_date) queryParams.append('start_date', params.start_date);
    if (params.end_date) queryParams.append('end_date', params.end_date);
    if (params.search) queryParams.append('search', params.search);

    const queryString = queryParams.toString();
    const url = `/admin/audit-logs${queryString ? `?${queryString}` : ''}`;

    return apiClient.get<AuditLogListResponse>(url);
  },

  getStatistics: async (): Promise<AuditStatistics> => {
    return apiClient.get<AuditStatistics>('/admin/audit-logs/statistics');
  },

  getActions: async (): Promise<string[]> => {
    return apiClient.get<string[]>('/admin/audit-logs/actions');
  },

  getResources: async (): Promise<string[]> => {
    return apiClient.get<string[]>('/admin/audit-logs/resources');
  },

  getAuditLog: async (logId: string): Promise<AuditLog> => {
    return apiClient.get<AuditLog>(`/admin/audit-logs/${logId}`);
  },

  /**
   * Archive old audit logs for current tenant
   * Exports logs to JSON file, then deletes from database
   */
  archiveLogs: async (beforeDate?: string): Promise<ArchiveResult> => {
    const params = beforeDate ? `?before_date=${encodeURIComponent(beforeDate)}` : '';
    return apiClient.post<ArchiveResult>(`/admin/audit-logs/tenant/archive${params}`);
  },

  /**
   * List archived audit log files for current tenant
   */
  listArchives: async (): Promise<ArchiveListResponse> => {
    return apiClient.get<ArchiveListResponse>('/admin/audit-logs/tenant/archives');
  },

  /**
   * Get download URL for a tenant archive file
   */
  getArchiveDownloadUrl: (filename: string): string => {
    return `/api/v1/admin/audit-logs/tenant/archives/${encodeURIComponent(filename)}`;
  },
};
